File: packages/engine/src/memory/episodicMemory.ts
import { episodicEvents, getDrizzle } from 'db';
import { getEmbeddingModel } from '../llm/callLLMWithRetry';

export interface EpisodicEvent {
  id: string;
  event: string;
  importanceScore: number; // 1-10
  timestamp: Date;
  embedding?: number[];
  scope?: 'global' | 'tenant';
  businessId?: string;
}

export class EpisodicMemory {
  private userId: string;
  private businessId?: string;

  constructor(userId: string, businessId?: string) {
    this.userId = userId;
    this.businessId = businessId;
  }

  async addEvent(
    event: string,
    importanceScore: number,
    scope: 'global' | 'tenant' = 'tenant',
    businessId?: string,
  ): Promise<void> {
    if (!this.userId) {
      console.warn('[EpisodicMemory] Cannot add event without userId');
      return;
    }
    const targetBizId = scope === 'tenant' ? businessId || this.businessId || 'ecommerce' : null;

    console.log(
      `[EpisodicMemory] Added event for user ${this.userId} (Scope: ${scope}, Biz: ${targetBizId}): "${event}" with importance: ${importanceScore}`,
    );
    const embeddingModel = getEmbeddingModel();
    const embedding = await embeddingModel.embedQuery(event);

    const dbInstance = getDrizzle();
    if (dbInstance) {
      try {
        const serializedEmbedding = JSON.stringify(embedding);
        await dbInstance.insert(episodicEvents).values({
          userId: this.userId,
          businessId: targetBizId,
          scope: scope,
          content: event,
          embedding: serializedEmbedding,
          importance: importanceScore,
          timestamp: new Date(),
        });
        console.log(`[EpisodicMemory] Stored event directly in PostgreSQL [Scope: ${scope}]: "${event}"`);
        return;
      } catch (err) {
        console.warn('[EpisodicMemory] Drizzle insertion bypassed due to offline/failed DB.');
      }
    }
  }

  async retrieveEvents(query: string, limit = 3, precomputedEmbedding?: number[]): Promise<EpisodicEvent[]> {
    if (!this.userId) {
      return [];
    }
    console.log(`[EpisodicMemory] Retrieving episodic events for user ${this.userId} using query: ${query}`);

    let queryEmbedding = precomputedEmbedding;
    if (!queryEmbedding || queryEmbedding.length === 0) {
      const embeddingModel = getEmbeddingModel();
      queryEmbedding = await embeddingModel.embedQuery(query);
    }

    const dbInstance = getDrizzle();
    if (dbInstance) {
      try {
        const { eq, desc } = require('drizzle-orm');
        // 限制最多只读取最近的 50 条事件，避免历史会话过多时进行海量全表扫描与无谓的日志打印
        const allEvents = await dbInstance
          .select({
            id: episodicEvents.id,
            content: episodicEvents.content,
            importance: episodicEvents.importance,
            embedding: episodicEvents.embedding,
            timestamp: episodicEvents.timestamp,
            scope: episodicEvents.scope,
            businessId: episodicEvents.businessId,
          })
          .from(episodicEvents)
          .where(eq(episodicEvents.userId, this.userId))
          .orderBy(desc(episodicEvents.timestamp))
          .limit(50);

        // 🛡️ 多租户情景隔离与防投毒 (Dual-Tier Scoped Episodic Isolation)
        const tenantEvents = allEvents.filter((row: any) => {
          if (!row.scope || row.scope === 'global') {
            return true;
          }
          if (row.scope === 'tenant') {
            if (this.businessId && this.businessId !== 'ecommerce') {
              return row.businessId === this.businessId;
            }
            return !row.businessId || row.businessId === 'ecommerce' || row.businessId === this.businessId;
          }
          return false;
        });

        if (tenantEvents.length > 0) {
          const scoredEvents = tenantEvents.map((row: any) => {
            let embeddingArray: number[] | null = null;
            if (row.embedding) {
              try {
                embeddingArray = typeof row.embedding === 'string' ? JSON.parse(row.embedding) : row.embedding;
              } catch (e) {
                console.warn('[EpisodicMemory] Failed to parse embedding JSON:', e);
              }
            }

            let similarity = 0;
            if (embeddingArray && Array.isArray(embeddingArray) && embeddingArray.length === queryEmbedding.length) {
              // cosine similarity = (A . B) / (||A|| * ||B||)
              let dotProduct = 0;
              let normA = 0;
              let normB = 0;
              for (let i = 0; i < queryEmbedding.length; i++) {
                dotProduct += queryEmbedding[i] * embeddingArray[i];
                normA += queryEmbedding[i] * queryEmbedding[i];
                normB += embeddingArray[i] * embeddingArray[i];
              }
              const denominator = Math.sqrt(normA) * Math.sqrt(normB);
              similarity = denominator === 0 ? 0 : dotProduct / denominator;
            }

            // 结合关键词重合度做混合打分
            const queryTokens = query
              .toLowerCase()
              .split(/[\s,，、。!！?？]+/)
              .filter((t) => t.length >= 2);
            let keywordMatches = 0;
            for (const token of queryTokens) {
              if (row.content.toLowerCase().includes(token)) {
                keywordMatches++;
              }
            }
            const keywordScore = queryTokens.length > 0 ? (keywordMatches / queryTokens.length) * 0.95 : 0;
            const effectiveScore = Math.max(similarity, keywordScore);

            return {
              event: {
                id: row.id,
                event: row.content,
                importanceScore: row.importance || 3,
                timestamp: row.timestamp || new Date(),
                embedding: embeddingArray || undefined,
                scope: row.scope || 'global',
                businessId: row.businessId || undefined,
              },
              similarity: effectiveScore,
            };
          });

          // Sort descending by similarity
          scoredEvents.sort((a, b) => b.similarity - a.similarity);

          // 🔍 性能与 Prompt 优化：建立相似度硬阈值过滤（最低 0.55），
          // 剔除风马牛不相及的事件流落入 Prompt 造成上下文膨胀、延迟飙升
          const SIMILARITY_THRESHOLD = 0.55;
          const filteredEvents = scoredEvents.filter((item: any) => {
            const isPassed = item.similarity >= SIMILARITY_THRESHOLD;
            if (!isPassed && item.similarity > 0) {
              console.log(
                `[EpisodicMemory Filter] Event irrelevant, similarity ${item.similarity.toFixed(3)} below ${SIMILARITY_THRESHOLD}: "${item.event.event}"`,
              );
            }
            return isPassed;
          });

          // Limit and return top limit
          return filteredEvents.slice(0, limit).map((se: any) => se.event);
        }
      } catch (err) {
        console.warn('[EpisodicMemory] TS-based cosine similarity search bypassed due to offline/failed DB.', err);
      }
    }

    return [];
  }
}

File: services/engine-py/src/engine_py/memory/episodic_memory.py
"""情境记忆 — 镜像 memory/episodicMemory.ts(带时间戳事件 + 双层租户隔离)。"""

from __future__ import annotations

import json
import math
import re

from sqlalchemy import select

from ..db import EpisodicEventRow, get_session
from ..llm import get_embedding_model

_TOKEN_SPLIT_RE = re.compile(r"[\s,，、。!！?？]+")


def _cosine(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        return 0
    dot = sum(x * y for x, y in zip(a, b))
    norm = math.sqrt(sum(x * x for x in a)) * math.sqrt(sum(y * y for y in b))
    return dot / norm if norm else 0


def _parse_embedding(raw) -> list[float] | None:
    if not raw:
        return None
    try:
        value = json.loads(raw) if isinstance(raw, str) else raw
        return value if isinstance(value, list) else None
    except Exception:
        return None


class EpisodicMemory:
    def __init__(self, user_id: str, business_id: str | None = None) -> None:
        self.user_id = user_id
        self.business_id = business_id

    async def add_event(
        self,
        event: str,
        importance_score: int,
        scope: str = "tenant",
        business_id: str | None = None,
    ) -> None:
        if not self.user_id:
            print("[EpisodicMemory] Cannot add event without userId")
            return
        target_biz_id = business_id or self.business_id or "ecommerce" if scope == "tenant" else None
        embedding = await get_embedding_model().aembed_query(event)
        try:
            async with get_session() as session:
                session.add(
                    EpisodicEventRow(
                        user_id=self.user_id,
                        business_id=target_biz_id,
                        scope=scope,
                        content=event,
                        embedding=json.dumps(embedding),
                        importance=importance_score,
                    )
                )
                await session.commit()
        except Exception:
            print("[EpisodicMemory] Insertion bypassed due to offline/failed DB.")

    async def retrieve_events(
        self, query: str, limit: int = 3, precomputed_embedding: list[float] | None = None
    ) -> list[dict]:
        if not self.user_id:
            return []

        query_embedding = precomputed_embedding
        if not query_embedding:
            query_embedding = await get_embedding_model().aembed_query(query)

        try:
            async with get_session() as session:
                rows = (
                    (
                        await session.execute(
                            select(EpisodicEventRow)
                            .where(EpisodicEventRow.user_id == self.user_id)
                            .order_by(EpisodicEventRow.timestamp.desc())
                            .limit(50)
                        )
                    )
                    .scalars()
                    .all()
                )
        except Exception as err:
            print(f"[EpisodicMemory] cosine similarity search bypassed due to offline/failed DB: {err}")
            return []

        def _tenant_visible(row: EpisodicEventRow) -> bool:
            if not row.scope or row.scope == "global":
                return True
            if row.scope == "tenant":
                if self.business_id and self.business_id != "ecommerce":
                    return row.business_id == self.business_id
                return not row.business_id or row.business_id == "ecommerce" or row.business_id == self.business_id
            return False

        visible_rows = [row for row in rows if _tenant_visible(row)]
        if not visible_rows:
            return []

        query_tokens = [t for t in _TOKEN_SPLIT_RE.split(query.lower()) if len(t) >= 2]
        scored = []
        for row in visible_rows:
            embedding_array = _parse_embedding(row.embedding)
            similarity = _cosine(query_embedding, embedding_array) if embedding_array else 0
            content_lower = (row.content or "").lower()
            keyword_matches = sum(1 for token in query_tokens if token in content_lower)
            keyword_score = (keyword_matches / len(query_tokens)) * 0.95 if query_tokens else 0
            effective_score = max(similarity, keyword_score)
            scored.append(
                {
                    "event": {
                        "id": str(row.id),
                        "event": row.content,
                        "importanceScore": row.importance or 3,
                        "timestamp": (row.timestamp or "").isoformat() if row.timestamp else None,
                        "embedding": embedding_array or None,
                        "scope": row.scope or "global",
                        "businessId": row.business_id or None,
                    },
                    "similarity": effective_score,
                }
            )

        scored.sort(key=lambda item: item["similarity"], reverse=True)
        threshold = 0.55
        filtered = [item for item in scored if item["similarity"] >= threshold]
        return [item["event"] for item in filtered[:limit]]
